src/mdns/mdns_mdnssd.cpp

```diff
@@ -883,57 +895,73 @@ void PublisherMDnsSd::ServiceSubscription::HandleBrowseResult(DNSServiceRef
     VerifyOrExit(aErrorCode == kDNSServiceErr_NoError);
     VerifyOrExit(aFlags & kDNSServiceFlagsAdd);
 
-    DeallocateServiceRef();
     Resolve(aInterfaceIndex, aInstanceName, aType, aDomain);
 
 exit:
     if (aErrorCode != kDNSServiceErr_NoError)
     {
-        mMDnsSd->OnServiceResolveFailed(*this, aErrorCode);
+        mMDnsSd->OnServiceResolveFailed(mType, mInstanceName, aErrorCode);
+        DeallocateServiceRef();
     }
 }
 
 void PublisherMDnsSd::ServiceSubscription::Resolve(uint32_t           aInterfaceIndex,
                                                    const std::string &aInstanceName,
                                                    const std::string &aType,
                                                    const std::string &aDomain)
+{
+    mResolvingInstances.emplace_back(*this, aInstanceName, aType, aDomain, aInterfaceIndex);
+
+    mResolvingInstances.back().Resolve();
+}
+
+void PublisherMDnsSd::ServiceSubscription::RemoveInstanceResolution(
+    PublisherMDnsSd::ServiceInstanceResolution &aInstanceResolution)
+{
+    std::remove_if(
```

Comment:
`std::remove_if` doesn't really erase elements from the container. Instead, it moves removed elements to the end of the container.  You may want to `a.erase(std::remove_if(a.begin(), a.end(), ...), a.end())`.

Reply:
Looks like the predicate of `std::remove_if` is depending on the index of the target element. In that case I think it's safer to use `std::vector::erase` and `std::vector::find_if`. `std::remove_if` **might** cause issues if it changes elements' indexes during the look up.

Reply:
Fixed.